test/puppetlabs/puppetdb/mq_listener_test.clj

```diff
@@ -0,0 +1,125 @@
+(ns puppetlabs.puppetdb.mq-listener-test
+  (:import [java.util.concurrent TimeUnit])
+  (:require [clojure.test :refer :all]
+            [clojure.string :as str]
+            [clojure.tools.logging :refer [*logger-factory*]]
+            [puppetlabs.puppetdb.testutils :as tu]
+            [puppetlabs.puppetdb.mq-listener :refer :all]
+            [puppetlabs.trapperkeeper.testutils.logging :refer [atom-logger
+                                                                with-logging-to-atom
+                                                                with-log-suppressed-unless-notable]]
+            [puppetlabs.puppetdb.test-protocols :as test-protos]))
+
+(def max-log-check-attempts 1000)
+(def log-sleep-duration-in-ms 10)
+
+(defn await-log-entry
+  "This function looks for a log entry in `log-atom`. If one is not
+  found it sleeps for `log-sleep-duration-in-ms` and repeats the
+  check. It does this `max-log-check-attempts` times and will
+  eventually throw an exception if it's not found"
+  [log-atom]
+  (loop [times 0]
+    (when-not (seq @log-atom)
+      (if (> times max-log-check-attempts)
+        (-> (* log-sleep-duration-in-ms max-log-check-attempts)
+            (format "Log entry not found after %s ms")
+            RuntimeException.
+            throw)
+        (do
+          (Thread/sleep log-sleep-duration-in-ms)
+          (recur (inc times)))))))
+
+(def critical-error-pred (comp #{:fatal :error} :level))
+
+(deftest threadpool-logging
+  (testing "successful message"
+    (let [log-output (atom [])]
+      (with-log-suppressed-unless-notable critical-error-pred
+        (with-logging-to-atom "puppetlabs.puppetdb.mq-listener" log-output
+          (let [{:keys [threadpool semaphore] :as threadpool-ctx} (create-command-handler-threadpool 1)
+                handler-fn (tu/mock-fn)
+                identity-handler (command-submission-handler threadpool-ctx handler-fn)]
+            (try
+
+              (is (= 1 (.availablePermits semaphore)))
+              (is (not (test-protos/called? handler-fn)))
+
+              (identity-handler "this arg does nothing")
+
+              (is (.tryAcquire semaphore 1 TimeUnit/SECONDS)
```

Comment:
Might make this shorter in the next version - for this one test it's no big deal, but very many of these would add up?